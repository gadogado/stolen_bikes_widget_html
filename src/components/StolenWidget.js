import React, { Component } from 'react';
import Config from '../config';
import searchIcon from '../icons/search.svg';
import { fetchStolenNearby } from '../api';
import List from './List';
import '../styles/stolen-widget.scss';

const Loading = () => <div>Loading...</div>;

export default class StolenWidget extends Component {
  state = {
    results: [],
    loading: true,
    serialNumber: '',
  };

  async componentDidMount() {
    const { location = Config.defaultLocation } = this.props;
    const { bikes: results } = await fetchStolenNearby(location);
    this.setState({ loading: false, results });
  }

  searchSerial = async () => {
    const { serialNumber } = this.state;
    // TODO: search serial
  };

  onClickSearch = e => {
    e.preventDefault();
    this.searchSerial();
  };

  onSubmitSearch = e => {
    e.preventDefault();
    this.searchSerial();
  };

  onChangeSerial = e => {
    const serialNumber = e.target.value;
    this.setState({ serialNumber });
  };

  render() {
    const { loading, serialNumber, results } = this.state;
    return loading ? (
      <Loading />
    ) : (
      <div id="stolen-widget">
        <form className="topsearcher" onSubmit={this.onSubmitSearch}>
          <input
            type="text"
            placeholder="Search for a serial number"
            value={this.serialNumber}
            onChange={this.onChangeSerial}
          />
          <a href="#" className="subm" onClick={this.onClickSearch}>
            <img src={searchIcon} />
          </a>
        </form>
        <List results={results} />
      </div>
    );
  }
}
