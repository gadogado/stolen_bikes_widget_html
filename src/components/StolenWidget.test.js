import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import StolenWidget from './StolenWidget';

describe('<StolenWidget /> shallow rendered', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<StolenWidget />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('accepts config props', () => {
    const wrap = shallow(
      <StolenWidget
        cacheResults
        recentResults={false}
        location="Portland, OR"
        height={400}
      />,
    );
    const { props } = wrap.instance();
    const expected = {
      cacheResults: true,
      recentResults: false,
      location: 'Portland, OR',
      height: 400,
    };
    expect(props).toEqual(expected);
  });

  it('is loading by default', () => {
    const wrap = shallow(<StolenWidget />);
    expect(wrap.state('loading')).toEqual(true);
  });

  it('has a disabled search input', () => {
    const wrap = shallow(<StolenWidget />);
    expect(wrap.find('.topsearcher input').is('[disabled]')).toBe(true);
  });

  it('on search input changes searchToken', () => {
    const wrap = shallow(<StolenWidget />);
    const input = wrap.find('input');
    input.simulate('change', { target: { value: 'something' } });
    expect(wrap.state('searchToken')).toEqual('something');
  });

  it('calls searchSerial when the search button is clicked', () => {
    const wrap = shallow(<StolenWidget />);
    const event = Object.assign(jest.fn(), { preventDefault: () => {} });
    const instance = wrap.instance();

    jest.spyOn(instance, 'searchSerial');
    wrap.find('button').simulate('click', event);
    expect(instance.searchSerial).toHaveBeenCalled();
  });

  describe.skip('<StolenWidget /> mount', () => {
    it.skip('has a disabled loading state after componentDidMount');
    it.skip('has a list of recent stolen bikes');
    it.skip('has a list of bikes with matching serial numbers');
    it.skip('adjusts the height of the widget');
    it.skip('displays a message for no results');
  });
});
