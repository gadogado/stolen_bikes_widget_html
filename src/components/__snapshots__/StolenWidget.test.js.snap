// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StolenWidget /> shallow rendered matches the snapshot 1`] = `
<div
  id="stolen-widget"
>
  <form
    className="topsearcher"
    onSubmit={[Function]}
  >
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Search for a serial number"
      type="text"
      value=""
    />
    <button
      className="subm"
      onClick={[Function]}
      type="submit"
    >
      <img
        alt="search"
        src="test-file-stub"
      />
    </button>
  </form>
  <div
    className="binxcontainer"
    id="binx_list_container"
  >
    <Loading />
  </div>
</div>
`;
